vscode/src/circuit.ts

```diff
@@ -0,0 +1,172 @@
+      reason: "timeout",
+      flowStatus: UserFlowStatus.Aborted,
+    });
+    log.info("terminating circuit worker due to timeout");
+    worker.terminate();
+  }, compilerRunTimeoutMs);
+  let title;
+  let subtitle;
+  const targetProfile = getTarget();
+  const sources = await loadProject(editor.document.uri);
+  if (operation) {
+    title = `${operation.operation} with ${operation.totalNumQubits} input qubits`;
+    subtitle = `${getTargetFriendlyName(targetProfile)} `;
+  } else {
+    title = basename(editor.document.uri.path) || "Circuit";
+    subtitle = `${getTargetFriendlyName(targetProfile)}`;
+  }
+
+  try {
+    sendTelemetryEvent(EventType.CircuitStart, { associationId }, {});
+    const circuit = await worker.getCircuit(sources, targetProfile, operation);
+    clearTimeout(compilerTimeout);
+
+    const message = {
+      command: "circuit",
+      circuit,
+      title,
+      subtitle,
+    };
+    sendMessageToPanel("circuit", false, message);
+
+    sendTelemetryEvent(EventType.CircuitEnd, {
+      associationId,
+      flowStatus: UserFlowStatus.Succeeded,
+    });
+  } catch (e: any) {
+    log.error("Circuit error. ", e.toString());
+    clearTimeout(compilerTimeout);
+
+    const errors: [string, VSDiagnostic][] =
+      typeof e === "string" ? JSON.parse(e) : undefined;
+    let errorHtml = "There was an error generating the circuit.";
+    if (errors) {
+      errorHtml = errorsToHtml(errors);
+    }
+
+    if (!timeout) {
+      sendTelemetryEvent(EventType.CircuitEnd, {
+        associationId,
+        reason: errors && errors[0] ? errors[0][1].code : undefined,
+        flowStatus: UserFlowStatus.Failed,
+      });
+    }
+
+    const message = {
+      command: "circuit",
+      title,
+      subtitle,
+      errorHtml,
+    };
+    sendMessageToPanel("circuit", false, message);
+  } finally {
+    log.info("terminating circuit worker");
+    worker.terminate();
+  }
+}
+
+/**
+ * Formats an array of compiler/runtime errors into HTML to be presented to the user.
+ *
+ * @param {[string, VSDiagnostic][]} errors
+ *  The string is the document URI or "<project>" if the error isn't associated with a specific document.
+ *  The VSDiagnostic is the error information.
+ *
+ * @returns {string} - The HTML formatted errors, to be set as the inner contents of a container element.
+ */
+function errorsToHtml(errors: [string, VSDiagnostic][]): string {
+  let errorHtml = "";
+  for (const error of errors) {
+    let location;
+    const document = error[0];
+    try {
+      // If the error location is a document URI, create a link to that document.
+      // We use the `vscode.open` command (https://code.visualstudio.com/api/references/commands#commands)
+      // to open the document in the editor.
+      // The line and column information is displayed, but are not part of the link.
+      //
+      // At the time of writing this is the only way we know to create a direct
+      // link to a Q# document from a Web View.
```

Comment:
A WebView can also send a message back to the main thread, which can then take whatever action the extension can take. We don't handle any custom messages currently, but I put the infrastructure in (see webviewPanel.ts@504)

Reply:
Oh, that's potentially even better... is it ok if I just update this comment and worry about it later? The current link approach seemed like the easiest way to get the job done.